core/src/main/scala/kafka/server/DelayedCreateTopics.scala

```diff
@@ -0,0 +1,91 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import kafka.api.LeaderAndIsr
+import kafka.common.TopicAndPartition
+import org.apache.kafka.common.protocol.Errors
+
+import scala.collection._
+
+/**
+  * The create metadata maintained by the delayed create operation
+  *
+  * TODO: local state doesn't count, need to know state of all relevant brokers
+  *
+  */
+case class CreateTopicMetadata(topic: String, replicaAssignments: Map[Int, Seq[Int]], error: Errors)
+
+/**
+  * A delayed create topics operation that can be created by the admin manager and watched
+  * in the topic purgatory
+  */
+class DelayedCreateTopics(delayMs: Long,
+                          createMetadata: Seq[CreateTopicMetadata],
+                          adminManager: AdminManager,
+                          responseCallback: Map[String, Errors] => Unit)
+  extends DelayedOperation(delayMs) {
+
+  /**
+    * The operation can be completed if all of the topics exist every partition has a leader in the controller.
+    * See KafkaController.onNewTopicCreation
+    */
+  override def tryComplete() : Boolean = {
+    trace(s"Trying to complete operation for $createMetadata")
+
+    val leaderlessPartitionCount = createMetadata.foldLeft(0) { case (topicCounter, metadata) =>
```

Comment:
The reason this was not blocking for a long time (I would have seen it in the tests) is because whenever there is an error we throw and exception and set the metadata to: 
```scala
CreateTopicMetadata(topic, Map(), Errors.forException(e))
```

That Map() is the assignments, since there are no assignments it wont wait for the creation in the delayedProduce. 

That said I think an explicit error filter should be used here too, just to be sure.